overcome limitation of 100 LogStreamNames

vlinevych opened the pull request:
`rds_exporter` crashes when there are more than 100 instances configured for enhanced monitoring.
The issue also exists in the upstream project.

RDS exporter calls cloudwatch's method `FilterLogEvents` with a parameter `LogStreamNames`,
which cannot be longer than 100 elements:
https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_FilterLogEvents.html

This PR works it around by calling FilterLogEvents multiple times.

The changes are at lines 67-76. The rest is just an indentation required to nest the logic into the loop.

mauipipe commented on ahoy/rds-exporter/custom-templates/prometheus-servicemonitor.yaml:
```diff
@@ -43,4 +43,4 @@ spec:
       app: {{ template "rds-exporter.name" $ }}-{{ $key }}
 {{- end }}
-{{- end }}
+{{- end }}
```
This file should be already present in the template. Without removing the template folder the plugin will not be used

spetrashov commented:
+1

mauipipe commented:
+1

RTodorov commented:
+1

Commit:
overcome limitation of 100 LogStreamNames

karakanb commented on enhanced/scraper.go:
```diff
@@ -59,60 +59,72 @@ func (s *scraper) start(ctx context.Context, interval time.Duration, ch chan<- m
+	// https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_FilterLogEvents.html
+	for i := 0; len(s.logStreamNames)-(100*i) > 0; i++ {
+		slice_start := i * 100
+		slice_end := slice_start + 100
```
You might want to use `camelCase` naming for consistency.
```suggestion
		sliceStart := i * 100
		sliceEnd := sliceStart + 100
```

karakanb commented on enhanced/scraper.go:
```diff
@@ -59,60 +59,72 @@ func (s *scraper) start(ctx context.Context, interval time.Duration, ch chan<- m
+
+	// LogStreamNames parameter supports up to 100 items.
+	// https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_FilterLogEvents.html
+	for i := 0; len(s.logStreamNames)-(100*i) > 0; i++ {
```
you might want to move this `len` call to a separate variable just to be safe, the compiler might not be able to optimize this, and if it cannot you'd be counting the stream on every iteration of the loop. I guess a simpler version of it would be something along the lines of this:

```suggestion
	streamCount := len(s.logStreamNames)
	batchSize := 100
	batchCount := (streamCount / batchSize) + 1
	for i := 0; i < batchCount; i++ {
```

puellanivis replied:
Better to just do `i := 0; i < len(s.logStreamNames; i += 100 { … }` and let the computer do the maths for you. 😉

puellanivis replied:
Oh, as an aside, unlike some languages, in Go the `len()` function does not actually count the elements of a slice in `O(n)` time. It just reads the `Len` field of the slice header: https://golang.org/pkg/reflect/#SliceHeader which is `O(1)` time.

puellanivis commented on enhanced/scraper.go:
```diff
@@ -59,60 +59,72 @@ func (s *scraper) start(ctx context.Context, interval time.Duration, ch chan<- m
+
+	// LogStreamNames parameter supports up to 100 items.
+	// https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_FilterLogEvents.html
+	for i := 0; len(s.logStreamNames)-(100*i) > 0; i++ {
```
Consider instead: `for i := 0; i < len(s.logStreamNames; i += 100 { … }`

karakanb reviewed (commented):
I suggest moving the core of this loop to a new function, which would allow you to pass the names of the resources you'd like to fetch, and you can easily span goroutines to parallelize them. 

A very simple example of parallelizing this can be as follows, just to give an idea:
```go
// replace this one with the function that actually fetches the received metrics.
func fetchLogs(items []int) {
	fmt.Println(items)
}

// and this demonstrates a part of your existing scrape function
func main() {
	batchCount := 5
	
	// think of this one as a synchronizer item that holds a simple counter.
	var waitGroup sync.WaitGroup
	
	// this should wait until its internal counter hits the batch count.
	waitGroup.Add(batchCount)

	for i := 0; i < batchCount; i++ {
		// things will run concurrently when they start with `go`.
		go func(items []int) {
			
			// do the actual operation.
			fetchLogs(items)
			
			// increment the internal counter of the synchronizer value.
			waitGroup.Done()
		}([]int{i, i + 1, i + 2})
	}

	// wait until your internal counter has reached its expected value.
	waitGroup.Wait()
	
	// that's it, move on with your life.
	fmt.Println("completed")
}
```

I don't know how familiar you are with Go, so I just wanted to give a quick example of it.

puellanivis commented on enhanced/scraper.go:
```diff
@@ -59,60 +59,72 @@ func (s *scraper) start(ctx context.Context, interval time.Duration, ch chan<- m
+			LogStreamNames: aws.StringSlice(s.logStreamNames[slice_start:slice_end]),
+			StartTime:      aws.Int64(aws.TimeUnixMilli(s.nextStartTime)),
+		}
+		s.logger.Debugf("Requesting metrics since %s (last %s).", s.nextStartTime.UTC(), time.Since(s.nextStartTime))
```
You can use the rich-logging of fields instead: `s.logger.With("next_start", s.nextStartTime).With("since_last", time.Since(s.nextStartTime)).Debugf("requesting metrics")`

Rather than composing a `printf` string yourself.

puellanivis commented on enhanced/scraper.go:
```diff
@@ -59,60 +59,72 @@ func (s *scraper) start(ctx context.Context, interval time.Duration, ch chan<- m
+					s.logger.Warnf("Enhanced monitoring problem: %s", awsErr.Message())
+				}
+			} else {
+				s.logger.Errorf("Failed to filter log events: %s.", err)
```
Seems like the period at the end is there in the original, but Golang prefers log messages to not end with punctuation (they might be expanded on or added to, so using punctuation is not recommended at their end).

puellanivis commented:
I highly recommend not making this parallelized execution until you are way more familiar with Go. Learn to crawl before you try and fly. Parallelization is way too easy to get subtly wrong.

puellanivis commented on enhanced/scraper.go:
```diff
@@ -59,60 +59,72 @@ func (s *scraper) start(ctx context.Context, interval time.Duration, ch chan<- m
 
-			if allMetrics[instance.ResourceID] == nil {
-				allMetrics[instance.ResourceID] = make(map[time.Time][]prometheus.Metric)
+				if allMetrics[instance.ResourceID] == nil {
+					allMetrics[instance.ResourceID] = make(map[time.Time][]prometheus.Metric)
+				}
+				timestamp := aws.MillisecondsTimeValue(event.Timestamp)
+				metrics := osMetrics.makePrometheusMetrics(instance.Region)
+				allMetrics[instance.ResourceID][timestamp] = metrics
```
About the “do not paralelize this code”, in particular all of this code would all need to be done holding a mutex lock.

It is way better to just do things synchronously, until and unless you have profiling information showing that running this synchronously is performance bottleneck.

Commit:
Adressed PR comments:
- loop parameters are more readable
- use rich logger instead of Debugf
- no punctuation at the end of the log message
- camelCase

karakanb commented:
+1

![Approved with Zappr](https://img.shields.io/badge/dev-%E2%9C%93-brightgreen.svg?style=flat-square)  ![Approved with Zappr](https://img.shields.io/badge/dev_qa-%E2%9C%93-brightgreen.svg?style=flat-square)

puellanivis commented:
+1

![Approved with Zappr](https://img.shields.io/badge/dev-%E2%9C%93-brightgreen.svg?style=flat-square)  ![Approved with Zappr](https://img.shields.io/badge/dev_qa-%E2%9C%93-brightgreen.svg?style=flat-square)

craig-foster commented:
+1

![Approved with Zappr](https://img.shields.io/badge/pm-%E2%9C%93-brightgreen.svg?style=flat-square)

Validator2008 commented:
+1

![Approved with Zappr](https://img.shields.io/badge/access-%E2%9C%93-brightgreen.svg?style=flat-square)  ![Approved with Zappr](https://img.shields.io/badge/dev-%E2%9C%93-brightgreen.svg?style=flat-square)  ![Approved with Zappr](https://img.shields.io/badge/dev_qa-%E2%9C%93-brightgreen.svg?style=flat-square)  ![Approved with Zappr](https://img.shields.io/badge/pm-%E2%9C%93-brightgreen.svg?style=flat-square)  ![Approved with Zappr](https://img.shields.io/badge/qa-%E2%9C%93-brightgreen.svg?style=flat-square)  ![Approved with Zappr](https://img.shields.io/badge/tech_lead-%E2%9C%93-brightgreen.svg?style=flat-square)
